core_java/grep/README.md

```diff
@@ -0,0 +1,49 @@
+
+
+## Usage
+To use this application, we need to specify three arguments described below:
+* _regex_ - Regular expression which will be searched for in all files in the specified directory.
+* _rootPath_ - Tt is a path to the directory that will be searched recursively. All the sub directories of this directory will also be searched.
+* _outFile_ - Output file where all the matched lines will be saved.
+For example - "SampleRegex" is the text that needs to be searched. The app would search the *rootPath*, in this case, `home/centos/dev`  directory for any lines that contain text SampleRegex (__regex - .*SampleText.*__). Once the matches are found, it would write the list of matched lines to *OutFile*, in this case, `home\centos\grep.out` 
+`.*SampleRegex.* home/centos/dev home/centos/grep.out`
+
+## Pseudocode
+High level description of working of JavaGrep functions:
+
+```
+matchedLines = []
+for file in listFilesRecursively(rootDir)
+        for line in readLines(file)
+        if containsPattern(line)
+                matchedLines.add(line)
+        writeToFile(matchedLines)
+```
+
+
+Both of our implementations - JavaGrepImp and JavaGrepLambda uses JavaGrep function.
+We have implemented the above workflow in process method().
+Description of methods in the workfow:
+* _listFilesRecursively()_ - traverse the specified root directory recursively and generate list of files in the directory.
+* _readLines()_ - takes the 'file' and add all the generated lines to the file.
+* _containsPattern()_ - verifies if each line in the given file contains the regular expression specified by  argument.
+* _writeToFile()_ - If regex is matched, it writes those lines whose pattern is matched to a new output file specified by the user.
+
+
+## Performance Issue
+Our program is good enough for small files and will work efficiently. But on large files, it will work but not very efficiently. The problem with this method is all file lines are kept in the memory - which will quickly lead to OutOfMemoryError if the file is large, say 50GB. The reason is lines of files are stored in the memory. By keeping contents of file in the memory will quickly exhaust the available memory.
+1. What we need here is to just be able to iterate through each line of the file, do some processing and let it go, without keeping it in memory. In short, we need to iterate through the lines without holding the memory. This is called Streaming through the file. This solution will iterate through all the lines and process each line, without keeping any references in the memory. This way it consumes very less amount of memory. Hence, we have implemented this solution and we should not run into any memory issue with this implementation.
```

Comment:
Number is unordered!

Reply:
![1](https://user-images.githubusercontent.com/55670106/70594064-e17eee00-1ba4-11ea-84af-a4670401b56d.PNG)

I don't see any unordered number. Under **Performance Issue**, there are 3 points - 1,2,3
and under **Improvements**, there are 3 points 1,2,3.